lib/log.c

```diff
@@ -766,42 +775,55 @@ void closezlog(void)
 
 	XFREE(MTYPE_ZLOG, zl);
 	zlog_default = NULL;
+	pthread_mutex_unlock(&loglock);
 }
 
 /* Called from command.c. */
 void zlog_set_level(zlog_dest_t dest, int log_level)
 {
+	pthread_mutex_lock(&loglock);
 	zlog_default->maxlvl[dest] = log_level;
+	pthread_mutex_unlock(&loglock);
 }
 
 int zlog_set_file(const char *filename, int log_level)
 {
-	struct zlog *zl = zlog_default;
+	struct zlog *zl;
 	FILE *fp;
 	mode_t oldumask;
+	int ret = 1;
 
 	/* There is opend file.  */
 	zlog_reset_file();
 
+	pthread_mutex_lock(&loglock);
+
+	zl = zlog_default;
```

Comment:
the lock and pointer caching should be move to be within the 'set flags' else case.

Reply:
agreed